src/unicode.rs

```diff
@@ -0,0 +1,5 @@
+mod std_case_mapping_iter;
+pub mod titlecase;
+mod ucd_generated_case_mapping;
+
+pub use titlecase::{to_titlecase, Titlecase, ToTitlecase};
```

Comment:
But `Vec` isn't being used in `src/unicode/*` except in `tests`. I assume `cargo test --no-default-features` will catch this if `Vec` is actually used.
<details><summary>example</summary>
<img width="840" alt="image" src="https://github.com/artichoke/roe/assets/12410942/5fcff985-de70-4a1f-aac8-748f3bbfdb9e">
</details>

Reply:
Oh wow, good call! I did the naive thing of checking the ascii impl, and saw `to_ascii_lowercase` had the feature alloc around it.

The good news, after changing around the module structure, I now see what's going on I think.

It looks like `roe` currently exports not only the titlecase/lowercase/uppercase iterators, but also encoding specific functions - specifically things like `ascii::lowercase::to_ascii_lowercase`. I assume these functions are to be consumed by other Ruby string functions as short circuits for perf reasons (instead of slice => iterator => vec), and keeping all "case folding logic" in one crate.

This makes me wonder though, should `unicode::*` be exported or not. If yes, I think it would make sense to make `unicode::titlecase::to_titlecase` to be consistent with `ascii::titlecase::to_titlecase`, and thus to take a slice, and return a newly allocated `Vec<u8>`.

My preference:
- If we export `ascii::*` related methods, then we should also export `unicode::*` related things too
- `ascii::*` and `unicode::*` should take the same sort of parameters in this case

On "whether we should export `ascii::*`", I think it makes sense, and thus use `roe::ascii::*` instead of `std::str::*` in `artichoke-backend` and friends. @lopopolo thoughts?

Reply:
Note: if we do want to export `unicode::*`, then I would say for this PR we just don't export `unicode::*` for now, and then in the next PR do the work to bring some of the unicode functions into the `unicode::*` module etc.